docs/TypedMongo.md

```diff
@@ -0,0 +1,446 @@
+  SimpleEntity(ObjectId.get(), 3, "third")
+)
+
+import monix.eval.Task
+
+// every DB operation returns a Monix Task or Observable
+val task: Task[Unit] =
+  collection.insertMany(entities)
+
+// use whatever Scheduler is appropriate in your context (like ExecutionContext for Futures)
+import monix.execution.Scheduler.Implicits.global
+// run the Task
+task.foreach(_ => println("insert successful"))
+```
+
+### Finding documents by query
+
+```scala
+import monix.eval.Task
+import monix.reactive.Observable
+
+val observable: Observable[SimpleEntity] =
+  collection.find(SimpleEntity.ref(_.int) > 1)
+
+// use whatever Scheduler is appropriate in your context (like ExecutionContext for Futures)
+import monix.execution.Scheduler.Implicits.global
+// run the Observable
+observable.foreach(entity => println(s"Found entity: $entity"))
+
+// alternatively, collect all found entities into a List
+val listTask: Task[List[SimpleEntity]] = observable.toListL
+listTask.foreach(foundEntities => println(s"Found entities: $foundEntities"))
+```
+
+## Modeling entities
+
+A MongoDB entity may be:
+* a case class
+* a sealed trait/class with `@flatten` annotation (see `GenCodec`'s [flat format](GenCodec.md#flat-format) for more details)
+
+Also, every MongoDB entity must extend `MongoEntity` in order to specify the ID field and type. 
+Its companion must also extend `MongoEntityCompanion`. This causes two typeclass instances (implicits) 
+to be materialized for the entity type:
+
+* `GenCodec` - used for serialization of the entity into BSON. See [`GenCodec`](GenCodec.md) for more details.
+* `MongoAdtFormat` - captures the internal structure of the entity, making it possible to refer to properties of this
+  entity in order to build queries, updates, sort orders, etc.
+
+### Field types
+
+Any type that has a `GenCodec` instance will be accepted as a field type in MongoDB entity.
+However, the MongoDB API is aware of internal structure of some types, including:
+
+* [embedded documents](#embedded-document-types) - serialized into BSON documents
+* collections, i.e. any subtype of `scala.collection.Seq` or `scala.collection.Set` - serialized into BSON arrays
+* maps, i.e. any subtype of `scala.collection.Map` - serialized into BSON documents
+* option-like types, i.e. `Option`, `Opt`, `OptArg`, etc. - serialized into nullable values
+
+Being aware of internal structure makes it possible to build queries, updates, projections, etc. that reach inside
+these data types. For example, you can refer to a specific map entry in a query.
+
+### Embedded document types
+
+A type representing embedded document is very similar to an entity type, i.e. it must be:
+
+* a case class
+* a sealed trait/class with `@flatten` annotation
+
+However, an embedded document is not a toplevel entity and does not need an ID. Therefore, it doesn't need to
+extend `MongoEntity` and its companion must extend `MongoDataCompanion` rather than `MongoEntityCompanion`.
```

Comment:
Sometimes you really want to use an external type (even though it may be considered a bad practice). If it's possible to do that with this driver, maybe we could be less strict with the `must` and mention how to do this neatly.

Reply:
You want to use an external type as an _entity_ type or just _field_ type? As explained in this doc, you can use any type that has `GenCodec` as a field type (opaque).

Reply:
In my case it's field type (I also consider e.g. cross-compiled classes as external in this case). Just wanted to clarify that `must extend` is not really true.

Reply:
`must extend` is true for an _embedded document_ whose internal structure is transparent. This doc section is about these.